atc/worker/client.go

```diff
@@ -156,6 +173,62 @@ func (client *client) CreateVolume(logger lager.Logger, volumeSpec VolumeSpec, w
 	return worker.CreateVolume(logger, volumeSpec, workerSpec.TeamID, volumeType)
 }
 
+var checkProcessSpec = runtime.ProcessSpec{
+	Path: "/opt/resource/check",
+}
+
+func (client *client) RunCheckStep(
+	ctx context.Context,
+	logger lager.Logger,
+	owner db.ContainerOwner,
+	containerSpec ContainerSpec,
+	workerSpec WorkerSpec,
+	strategy ContainerPlacementStrategy,
+	containerMetadata db.ContainerMetadata,
+	resourceTypes atc.VersionedResourceTypes,
+	timeout time.Duration,
+	checkable resource.Resource,
+) (CheckResult, error) {
+	chosenWorker, err := client.pool.FindOrChooseWorkerForContainer(
+		ctx,
+		logger,
+		owner,
+		containerSpec,
+		workerSpec,
+		strategy,
+	)
+	if err != nil {
+		return CheckResult{}, fmt.Errorf("find or choose worker for container: %w", err)
+	}
+
+	container, err := chosenWorker.FindOrCreateContainer(
+		ctx,
+		logger,
+		&NoopImageFetchingDelegate{},
+		owner,
+		containerMetadata,
+		containerSpec,
+		resourceTypes,
+	)
+	if err != nil {
+		return CheckResult{}, fmt.Errorf("find or create container: %w", err)
+	}
+
+	deadline, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	versions, err := checkable.Check(deadline, checkProcessSpec, container)
+	if err != nil {
+		if err == context.DeadlineExceeded {
+			return CheckResult{}, fmt.Errorf("timed out after %v checking for new versions", timeout)
+		}
+
+		return CheckResult{}, fmt.Errorf("check: %w", err)
+	}
+
+	return CheckResult{Versions: versions}, nil
```

Comment:
Can we give the CheckResult an ExitStatus similar to:
https://github.com/concourse/concourse/blob/aac7b2f461ed2a3ad6fa8d861d439d3774f9fde8/atc/worker/client.go#L553-L560

Reply:
Yes, please.
This allows us to differentiate script errors vs. Concourse errors

Reply:
I'm a bit confused with how we should proceed here :thinking:

I see (and agree) that we want to be able to differentiate system errors from
executables returning a non-0 exit status, but, it seems like the current
behavior of "check error reporting" is tied to returning an error with that
stuff embedded into it, and I think changing that wouldn't be the scope of this
PR (which was meant to be structural only).

I say that because there seems to be a gap between the way that errors are
reported for `check`s and `get`s`/`put`s/`task`s:

- `check`s report *how* they errored through errors, being even recorded in the
  database under the `checks` table - the "reason" for the non-0 exit status is
  embedded in the error message from `runtime.ErrResourceScriptFailed`

https://github.com/concourse/concourse/blob/1ae69250ceef7480f56732377478c46642c8791a/atc/runtime/errors.go#L25-L31

- `get`s, `put`s, and `task`s have their "reason for a non-0" reported through
  their corresponding delegates: one can tell *how* a `get` failed by looking at
  the build logs, with `stderr` being directly tied to db events

wdyt?

thx!

Reply:
Yeah makes sense! 👍